### PR TITLE
MM-18919 Migrate 'utils/syntax_highlighting' to TypeScript

### PR DESCRIPTION
**Pull request by FlaviaBastos**
#### Summary
Migrates `utils/syntax_highlighting.jsx` to TypeScript.

#### Ticket Link
Fixes [issue #12484](https://github.com/mattermost/mattermost-server/issues/12484)

#### Note:
This is still a WIP and I need some direction, please.

`Constants.HighlightedLanguages` has an optional `aliases` key. 
I created a type for it in the `syntax_highlinting` file (see lines 128-134) but I'm not sure this is the correct approach. Also, because `aliases` can be optional, the type is showing as `string` or `undefined`, which throws an error for returning value when calling function `getLanguageFromNameOrAlias(language)` (line 168): `Type 'undefined' cannot be used as an index type.ts(2538)`, which doesn't have its return value properly typed yet.

As I'm not familiar with the whole codebase I hope some could please give some direction here, please.

### CODE REVIEW

File: utils/syntax_highlighting.tsx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    return null;
+}
+
+export function canHighlight(language: string): boolean {
+    return Boolean(getLanguageFromNameOrAlias(language));
+}
+
+export function getLanguageName(language: string): string {
+    if (canHighlight(language)) {
+        return HighlightedLanguages[getLanguageFromNameOrAlias(language)].name;
```
Comment (devinbinnie) [inline utils/syntax_highlighting.tsx]: Hey @FlaviaBastos, you should be able to do an explicit null check before this line, which should clear up the TS error. Something like: 
```suggestion
        const name: string | null = getLanguageFromNameOrAlias(language);
        if (!name) return '';
        return HighlightedLanguages[getLanguageFromNameOrAlias(language)].name;
```

Comment (FlaviaBastos) [reply utils/syntax_highlighting.tsx]: Great! Thanks for the suggestion @devinbinnie. 
I just pushed an update. Please let me know how it looks like.

**Comment by FlaviaBastos** [comment]
I see that the CI type-check is failing with a "Could not find a declaration file" for all files for the `highlight.js` library. I installed the `@types` package but I might have done something wrong. Any ideas, @devinbinnie?

**Comment by devinbinnie** [comment]
> I see that the CI type-check is failing with a "Could not find a declaration file" for all files for the `highlight.js` library. I installed the `@types` package but I might have done something wrong. Any ideas, @devinbinnie?

Weird, @jespino any ideas?

**Comment by hanzei** [comment]
@devinbinnie @jespino Do you have ideas how to fix this issue? https://github.com/mattermost/mattermost-webapp/pull/3862#issuecomment-539053693

**Comment by jespino** [comment]
@FlaviaBastos Sorry for the delay in my reply, I think we should follow the same approach here: https://stackoverflow.com/questions/54334893/load-only-highlight-js-library-using-typescript-not-default-with-all-languages using our own `.d.ts` files to define each language as a module.

**Comment by FlaviaBastos** [comment]
No worries, @jespino! Thanks for the pointer. I will work on that and reach out if I have questions. 
Thanks!

**Comment by FlaviaBastos** [comment]
Hi, @jespino! I just pushed an update with your suggestion and it seems to have resolved the issue. Thanks!

/cc @devinbinnie and @hanzei: 
Now the only issue is that my branch is way behind (I believe). Should I try to sync it myself? If so, if I try: `git fetch upstream` and `git pull upstream master`, it will probably add a new merge commit. Is this ok or do you have another workflow?

**Comment by jespino** [comment]
@FlaviaBastos Probably you only need to execute `git fetch upstream` and `git merge upstream/master` being in your PR branch, fix the conflicts, commint, and then `git push`

File: types/highlightjs.d.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+declare module 'highlight.js/lib/highlight.js';
```
Comment (jespino) [inline types/highlightjs.d.ts]: I think we don't need this concrete line, we should be using the proper types for that, the ones defined in the correct `@types` module, the problem was with the bindings, so all the other definitions are right.

File: utils/syntax_highlighting.tsx [hunk 3]
```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import hlJS from 'highlight.js/lib/highlight.js';
```
Comment (jespino) [inline utils/syntax_highlighting.tsx]: probably we need to import this like `import hlJS from 'highlight.js';` to makes the tipes work correctly.

File: utils/syntax_highlighting.tsx [hunk 4]
```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
```
Comment (jespino) [inline utils/syntax_highlighting.tsx]: I think the old ` utils/syntax_highlighting.jsx` should be removed.

**Comment by FlaviaBastos** [comment]
Thanks for all the feedback, @jespino! I'm in transit today and won't be able to address it right away but I will do as soon as I can and will let you know.

**Comment by jespino** [comment]
@FlaviaBastos Great, take your time, no rush :)

File: package.json [hunk 5]
```diff
@@ -7,6 +7,7 @@
   "private": true,
   "dependencies": {
     "@types/enzyme": "3.10.3",
+    "@types/highlight.js": "^9.12.3",
```
Comment (jespino) [inline package.json]: We normally use fixed versions in our dependencies, do you mind to remove the `^` here and, by the way, in the next line?

File: utils/syntax_highlighting.tsx [hunk 6]
```diff
@@ -1,8 +1,8 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
-import hlJS from 'highlight.js/lib/highlight.js';
-import hljsActionscript from 'highlight.js/lib/languages/actionscript.js';
+import hlJS from 'highlight.js';
+import hljsActionscript from 'highlight.js/lib/languages/actionscript.js'
```
Comment (jespino) [inline utils/syntax_highlighting.tsx]: Missing a semicolon at the end of this line.

**Comment by jespino** [comment]
@FlaviaBastos almost there, only a small set of changes and is ready for me.

File: package.json (hunk 5 above)
Comment (FlaviaBastos) [reply package.json]: Of course! 
(But the way, I'm SO glad you're online and already giving feedback! Sorry for the time it took me to get back to this. Also, this has been way more complicated than I anticipated, so I appreciate your patience)

File: utils/syntax_highlighting.tsx (hunk 6 above)
Comment (FlaviaBastos) [reply utils/syntax_highlighting.tsx]: Done

**Comment by devinbinnie** [comment]
@FlaviaBastos You have one lint error:
```
/root/mattermost-webapp/utils/syntax_highlighting.tsx
  176:10  error  Unexpected var, use let or const instead  no-var
```

**Comment by FlaviaBastos** [comment]
I think I fixed all errors now, @devinbinnie, but please let me know if there's anything else you would like to be changed.
Thanks!

**Pull request merged**